### PR TITLE
feat: implement asyncGenerator

### PR DESCRIPTION
**Pull request by gurgunday**
Would something like this speed up TTFB in certain scenarios?

I'll refactor the tests

### CODE REVIEW

**Review by mcollina — COMMENTED** [review]
You should test that the rendering is done in chunks.

File: src/html.js [hunk 1]
```diff
@@ -125,4 +125,92 @@ const htmlGenerator = function* ({ raw: literals }, ...expressions) {
...
+          if (expression.length) {
+            if (!isRaw) {
+              expression = expression.replace(escapeRegExp, escapeFunction);
+            }
+
+            yield expression;
+          }
+        }
+
+        continue;
```
Comment (mcollina) [inline src/html.js:194]: I think you should throw if this is the case

File: src/html.js [hunk 2]
```diff
@@ -125,4 +125,92 @@ const htmlGenerator = function* ({ raw: literals }, ...expressions) {
...
+          } else if (value === undefined || value === null) {
+            continue;
+          } else if (
+            typeof value[Symbol.iterator] === "function" ||
+            typeof value[Symbol.asyncIterator] === "function"
+          ) {
+            expression = "";
+
+            for await (const innerValue of value) {
+              expression += innerValue ?? "";
```
Comment (mcollina) [inline src/html.js:179]: You might need a couple of options about escaping here.

Comment (gurgunday) [reply src/html.js:179]: It uses the same logic, if that whole generator was prefixed with `!`, the whole thing is rendered raw, if not the whole thing is escaped

Also, you might've noticed but for consistency with `Array.join`, we don't take into account `undefined` and `null`

Comment (gurgunday) [reply src/html.js:179]: And btw, htmlGenerator will never return these types, only string, so these are here for consistency with the outer type handling

File: src/html.js [hunk 3]
```diff
@@ -125,4 +125,92 @@ const htmlGenerator = function* ({ raw: literals }, ...expressions) {
...
+          } else {
+            expression = `${value}`;
+          }
+
+          if (expression.length) {
+            if (!isRaw) {
+              expression = expression.replace(escapeRegExp, escapeFunction);
+            }
+
+            yield expression;
```
Comment (mcollina) [inline src/html.js:190]: This is problematic. You should be yielding things as they come inside the for..await loop (and escaping there, too)

Comment (gurgunday) [reply src/html.js:188]: I fixed it but the code is not looking good in my opinion, will refactor it a little more

File: src/html.js [hunk 4]
```diff
@@ -165,23 +151,35 @@ const htmlAsyncGenerator = async function* ({ raw: literals }, ...expressions) {
         }
 
         for await (const value of expressions[index]) {
-          if (typeof value === "string") {
-            expression = value;
-          } else if (value === undefined || value === null) {
+          if (value === undefined || value === null) {
             continue;
-          } else if (
+          }
```
Comment (mcollina) [inline src/html.js]: Can an expression be a `Promise`?

File: test/index.js [hunk 5]
```diff
@@ -163,38 +176,100 @@ test("htmlGenerator works with nested htmlGenerator calls in an array", () => {
...
   assert.strictEqual(
-    generator.next().value,
-    "<script>alert('This is an unsafe description.')</script>",
+    accumulator,
+    "<div><p>This is a safe description.<script>alert('This is an unsafe description.')</script>12345255</p></div>",
   );
-  assert.strictEqual(generator.next().value, "12345");
-  assert.strictEqual(generator.next().value, "255");
-  assert.strictEqual(generator.next().value, "</p>");
-  assert.strictEqual(generator.next().value, "</div>");
```
Comment (mcollina) [inline test/index.js:176]: You should have a test that verifies that X amount of chunks have been produced and not just one..

Comment (gurgunday) [reply test/index.js:176]: Just added [this](https://github.com/gurgunday/ghtml/pull/12/files#diff-5bb8db779819ddef5956a5d9d5949c05ef7445237656ca37bf2f02720271440bR266-R305), simulates a file being rendered for each element in an array, each call is a readFile promise, is it good?

File: src/html.js (hunk 4 above)
Comment (gurgunday) [reply src/html.js]: Passing `` htmlAsyncGenerator`${readFile('./staticPartial.html')}` `` is permitted, so yeah

**Comment by gurgunday** [comment]
My god, I'll add label or something for the benchmark, this is untenable 😄

**Comment by gurgunday** [comment]
But I'm thinking, instead of `for awaiting` in the inner loops, maybe we can batch them in an array

We'd still yield one by one in order as they get resolved but it could maybe improve performance

**Review by mcollina — APPROVED** [review]
lgtm, really nice

**Pull request merged**